numalogic/registry/dynamodb_registry.py

```diff
@@ -0,0 +1,398 @@
+
+        cached_artifact = self._load_from_cache(f"{part_key}::{v_sort_key}")
+
+        if cached_artifact:
+            _LOGGER.debug("Found cached artifact for key: %s", f"{part_key}::{v_sort_key}")
+            return cached_artifact
+
+        try:
+            response = table.get_item(Key={"skey": part_key, "dkey": v_sort_key})
+            item = response.get("Item") or {}
+        except ClientError as err:
+            raise DynamoDBRegistryError(f"{err.__class__.__name__} raised") from err
+
+        if item and "data" in item:
+            artifact_data = self._unpack_artifact_data(item.get("data"))
+            self._save_in_cache(f"{part_key}::{v_sort_key}", artifact_data)
+            return artifact_data
+
+        _LOGGER.info("Record not found for skey: %s, dkey: %s", part_key, sort_key)
+        return None
+
+    @staticmethod
+    def __save_item(
+        table, part_key: str, sort_key: str, data: dict[str, Any], version: Optional[int] = None
+    ):
+        try:
+            response = table.put_item(
+                Item={"skey": part_key, "dkey": sort_key, "data": data, "version": version}
+            )
+        except ClientError as err:
+            raise DynamoDBRegistryError(f"{err.__class__.__name__} raised") from err
+        return response
+
+    def save(
+        self,
+        skeys: KEYS,
+        dkeys: KEYS,
+        artifact: artifact_t,
+        **metadata: META_VT,
+    ) -> Optional[str]:
+        r"""
+        Saves the artifact into dynamodb.
+        Note that this overwrites any previously
+        saved item with the same skey and dkey.
+        Args:
+            artifact:
+            skeys: static key fields as list/tuple of strings
+            dkeys: dynamic key fields as list/tuple of strings
+            artifact: model to be saved
+            metadata: additional metadata surrounding the artifact that needs to be saved.
+        """
+        table = self.connector.get_resource().Table(self.table_name)
+        part_key = self.construct_key(skeys)
+        sort_key = self.construct_key(dkeys)
+
+        # Check if model exists
+        artifact_info = self.load(skeys, dkeys, latest=True)
+
+        if artifact_info:
+            version = int(artifact_info.extras.get("version")) + 1
+        else:
+            version = 1
+
+        artifact_data = self._pack_artifact_data(artifact, str(version), **metadata)
+
+        # Save a copy of the artifact with the current version.
+        self.__save_item(
+            table, part_key, self._construct_version_key(sort_key, version=version), artifact_data
+        )
+
+        # Save the zeroth version (v0) as the one pointing to the current version.
+        self.__save_item(
+            table, part_key, self._construct_version_key(sort_key), artifact_data, version=version
+        )
+
+        # Delete the stale model
+        stale_version = version - self.models_to_retain
+        if stale_version > 0:
```

Comment:
Question: Do you think we can delete models via TTL?

In the above logic if we fail to save v5 and v6 versions we will not be deleting v3 and v4 respectively.

Reply:
TTL can be enabled at any time on the DynamoDB table if needed as shown in the following document: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-how-to.html

Reply:
Yes. We can have that enabled.